Use the linker to check if a specific function is provided by the toolchain or runtime

nirbheek opened the pull request:
This allows build files to not bother with providing system headers under `prefix` while checking if a function exists, providing convenience similar to how autoconf macros work. This is especially convenient when you need to have multiple `#ifdef`s and `#define`s to get at the symbol in the header which would then have to be duplicated in the build file and in the source. It is also a much cleaner check since it ensures that the symbol is a function.

nirbheek commented:
The fallback of this is that if the header redefines the public symbol to something else, it won't be found. However, in that case the proper way is really to check if a particular header defines the symbol. I want to provide a patch for this, but I'm not sure what to name such a function. The best I can come up with is `cc.has_header_symbol()`. Suggestions are welcome. :)

tp-m commented:
I like ```cc.has_header_symbol()``` but maybe someone else can come up with something better :)

jpakkane commented:
Is this code taken from somewere else? The fact that the comment mentions HP-UX seems to indicate Autotools or something. Those are LGPL (I think) so we can't take code directly.

The GNU_SOURCE test might not pass on MSVC. I'll need to test that but don't have Windows access at the moment.

nirbheek commented:
Tim is on vacation for a few days so I'm answering on his behalf. The platform-specific information/reasoning was obtained from the autoconf implementation, but the code/comments were not copied directly afaik.

The `_GNU_SOURCE` test will get skipped on MSVC because `recvmmsg` isn't available there at all. Also, I forgot to mention that I have tested the patch itself on MSVC but not the test case yet. The `has_function` tests used by GStreamer had the same results before and after this patch.

jpakkane commented:
The GNU_SOURCE tests only check that either both of the symbols are found or neither is. Suppose there is a bug in the detector code which always returns false for the first check. In this case all tests would always pass even though the implementation is flawed.

Would it be possible to check this in a different way so this issue can't come up? Something like this perhaps:

    if host_machine.system() == 'linux'
      assert(cc.find_library(...))
    endif

nirbheek commented:
> Suppose there is a bug in the detector code which always returns false for the first check.

This is checked in the same test earlier: 

```
assert(cc.has_function('fprintf'), 'Existing function not found without include')
```

However, I think the test is indeed broken anyway since the `if` condition assumes that `has_function` behaves the old way. However, the `if linux` conditional might cause false alarms since not all versions of the kernel provide that syscall, and only glibc provides it afaik.

I think the proper way to do this test would be to change the conditional to `cc.has_header_symbol('sys/socket.h', 'recvmmsg', prefix : '#define _GNU_SOURCE')`. This, of course, means implementing that first, which we should do anyway I think since it's quite useful. Going to implement that next and then we can bikeshed on the name if we want to. ;)

nirbheek commented:
In the meantime, I cleaned up the comments in the `cc.has_function` implementation; they're much clearer now that we don't copy-paste them from the Autoconf macros (Tim clarified that the comments were indeed copied).

I also removed the hunk with the `"limits.h"` include which shouldn't be needed and added a new test which checks for the existence of `poll`, and tests that we can find it even if a provided header defines the symbol.

nirbheek commented:
`cc.has_header_symbol()` now exists and has tests, and `cc.has_function()` tests now use `has_header_symbol`. Going to test this on Windows/MSVC and Windows/MinGW soon.

jpakkane commented:
Could you explain what `has_header_symbol` does specifically and how it is different from the other tests. (This will probably be copypasted into the reference manual.)

nirbheek commented:
I just tested `cc.has_header_symbol` under both mingw and msvc on Windows and it all works fine. You have to use the `args` keyword to add the include path, of course, unless you're looking for a header inside the default include path.

`cc.has_header_symbol` has four purposes:

1) (This one can be copy-pasted into the manual)

`cc.has_header_symbol` allows one to detect whether a particular symbol (function, variable, #define, type definition, etc) is declared in the specified header.

The `prefix` keyword argument can be used to add `#define`s, `#include`s, etc that are required for the symbol to be declared (eg: `#define _GNU_SOURCE` is often required for some symbols to be exposed on Linux).

The `args` keyword argument can be used to pass a list of compiler arguments that are required to find the header or symbol. For example, you might need to pass the include path `-Isome/path/to/header` if the header is not in the default include path. Note that if you have a single prefix with all your dependencies, you might find it easier to append to the environment variables `C_INCLUDE_PATH` with gcc/clang and `INCLUDE` with msvc to expand the default include path.

2) The new `cc.has_function` cannot be used to detect functions provided by packages other than the compiler or the C runtime (libc on UNIX-like systems); which makes sense since the purpose of `has_function` is to check functions provided by the system, not by 3rd-party libraries.

3) `cc.has_function` could never be used to detect functions that are only declared in the header (inline functions, functions that are actually macros, etc).

4) `cc.has_function` cannot be used to find functions that are redefined by the header into some other symbol that can then be found in the standard library. I don't know of any examples of this (besides `__builtin_` in glibc/gcc which we support), but I'm sure some exist in some libc out there.

nioncode commented:
This broke test case 43 on Windows for all backends.

nirbheek commented:
What's the error for you? Can't give any ideas without that. ;) 

It works fine for me on Ninja + VS 2015, and Ninja + MinGW/GCC.

nioncode commented:
The first test for printf fails.

meson-log.txt:
> Build started at 2016-04-19T23:58:36.239406
> Python binary: C:\ProgramData\chocolatey\lib\python3\tools\python.exe
> Python system: Windows
> The Meson build system
> Version: 0.32.0.dev1
> Source dir: C:\meson\test cases\common\43 has function
> Build dir: C:\meson\pch
> Build type: native build
> Build machine cpu family: amd64
> Build machine cpu: amd64
> Project name: has function
> Native c compiler: cl (msvc 19.00.23506)
> Running compile:
> Command line:  cl C:\Users\vagrant\AppData\Local\Temp\tmpmyw5t79v.c kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /FoC:\Users\vagrant\AppData\Local\Temp\tmprl3irepb
> Code:
>  
>         #define printf meson_disable_define_of_printf
>         #include<stdio.h>
>         #undef printf
>         
>         #ifdef __cplusplus
>         extern "C"
>         #endif
>         char printf ();
>         
>         #if defined __stub_printf || defined __stub___printf
>         fail fail fail this function is not going to work
>         #endif
>         
>         int
>         main ()
>         {
>           return printf ();
>         }
> Compiler stdout:
>  Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23506 for x86
> 
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> 
> 
> 
> Compiler stderr:
>  tmpmyw5t79v.c
> 
> Microsoft (R) Incremental Linker Version 14.00.23506.0
> 
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> 
> 
> /out:tmprl3irepb.exe 
> 
> C:\Users\vagrant\AppData\Local\Temp\tmprl3irepb.obj 
> 
> kernel32.lib 
> 
> user32.lib 
> 
> gdi32.lib 
> 
> winspool.lib 
> 
> shell32.lib 
> 
> ole32.lib 
> 
> oleaut32.lib 
> 
> uuid.lib 
> 
> comdlg32.lib 
> 
> advapi32.lib 
> 
> tmprl3irepb.obj : error LNK2019: unresolved external symbol _printf referenced in function _main
> 
> LINK : MSIL .netmodule or module compiled with /GL found; restarting link with /LTCG; add /LTCG to the link command line to improve linker performance
> 
> Microsoft (R) Incremental Linker Version 14.00.23506.0
> 
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> 
> 
> /out:tmprl3irepb.exe 
> 
> C:\Users\vagrant\AppData\Local\Temp\tmprl3irepb.obj 
> 
> kernel32.lib 
> 
> user32.lib 
> 
> gdi32.lib 
> 
> winspool.lib 
> 
> shell32.lib 
> 
> ole32.lib 
> 
> oleaut32.lib 
> 
> uuid.lib 
> 
> comdlg32.lib 
> 
> advapi32.lib 
> 
> tmprl3irepb.obj : error LNK2001: unresolved external symbol _printf
> 
> tmprl3irepb.exe : fatal error LNK1120: 1 unresolved externals
> 
> 
> Running compile:
> Command line:  cl C:\Users\vagrant\AppData\Local\Temp\tmp2mm8vqfj.c kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /Od /FoC:\Users\vagrant\AppData\Local\Temp\tmpv1nh49po
> Code:
>  int main() { __builtin_printf; }
> Compiler stdout:
>  Microsoft (R) C/C++ Optimizing Compiler Version 19.00.23506 for x86
> 
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> 
> 
> 
> Compiler stderr:
>  tmp2mm8vqfj.c
> 
> C:\Users\vagrant\AppData\Local\Temp\tmp2mm8vqfj.c(1): error C2065: '__builtin_printf': undeclared identifier
> 
> 
> Checking for function "printf": NO
> 
> Meson encountered an error in file meson.build, line 6, column 2:
> Error encountered: Existing function not found.
>

nirbheek commented:
MSVC 2013 also has printf, so the problem isn't obvious. I tried installing MSVC 2013 but it failed because I don't have enough disk space on `C:` (installing on a separate drive doesn't actually use much space on that drive). Could you open an issue for this so it's not lost?
